### PR TITLE
Support groth16 implementation from ark

### PR DESCRIPTION
**Pull request by dark64**

### CODE REVIEW

File: changelogs/unreleased/1071-dark64 [hunk 1]
```diff
@@ -0,0 +1 @@
+Support for ark-groth16 implementation
```
Comment (Schaeff) [inline changelogs/unreleased/1071-dark64]:
```suggestion
Support for the groth16 scheme using the ark backend
```

File: zokrates_cli/src/helpers.rs [hunk 2]
```diff
@@ -93,16 +93,28 @@ impl TryFrom<(&str, &str, &str)> for Parameters {
...
-            (BackendParameter::Ark, CurveParameter::Bls12_377, SchemeParameter::MARLIN) => Ok(()),
+            (BackendParameter::Ark, CurveParameter::Bls12_377, SchemeParameter::GM17) => Ok(()),
+            #[cfg(feature = "ark")]
+            (BackendParameter::Ark, CurveParameter::Bls12_381, SchemeParameter::GM17) => Ok(()),
+            #[cfg(feature = "ark")]
+            (BackendParameter::Ark, CurveParameter::Bw6_761, SchemeParameter::GM17) => Ok(()),
             #[cfg(feature = "ark")]
             (BackendParameter::Ark, CurveParameter::Bn128, SchemeParameter::MARLIN) => Ok(()),
             #[cfg(feature = "ark")]
+            (BackendParameter::Ark, CurveParameter::Bls12_381, SchemeParameter::MARLIN) => Ok(()),
```
Comment (Schaeff) [inline zokrates_cli/src/helpers.rs:114]: Do I read correctly that bls12_381 was added for Marlin or is it a diff glitch? If so, it should be in the changelog

Comment (Schaeff) [reply zokrates_cli/src/helpers.rs:114]: Same for bls12_381 for gm17?

File: zokrates_cli/src/ops/generate_proof.rs [hunk 3]
```diff
@@ -112,20 +112,29 @@ pub fn exec(sub_matches: &ArgMatches) -> Result<(), String> {
...
         #[cfg(feature = "ark")]
+        Parameters(BackendParameter::Ark, _, SchemeParameter::G16) => match prog {
+            ProgEnum::Bn128Program(p) => cli_generate_proof::<_, _, G16, Ark>(p, sub_matches),
+            ProgEnum::Bls12_381Program(p) => cli_generate_proof::<_, _, G16, Ark>(p, sub_matches),
+            ProgEnum::Bls12_377Program(p) => cli_generate_proof::<_, _, G16, Ark>(p, sub_matches),
+            ProgEnum::Bw6_761Program(p) => cli_generate_proof::<_, _, G16, Ark>(p, sub_matches),
+        },
+        #[cfg(feature = "ark")]
         Parameters(BackendParameter::Ark, _, SchemeParameter::GM17) => match prog {
+            ProgEnum::Bn128Program(p) => cli_generate_proof::<_, _, GM17, Ark>(p, sub_matches),
```
Comment (Schaeff) [inline zokrates_cli/src/ops/generate_proof.rs:123]: Ah, here it is clearer :) so I guess updating the changelog is required

File: zokrates_core/src/proof_system/ark/gm17.rs [hunk 4]
```diff
@@ -1,43 +1,43 @@
...
 
 impl<T: Field + ArkFieldExtensions + NotBw6_761Field> NonUniversalBackend<T, GM17> for Ark {
     fn setup<I: IntoIterator<Item = Statement<T>>>(
         program: ProgIterator<T, I>,
     ) -> SetupKeypair<<GM17 as Scheme<T>>::VerificationKey> {
-        let parameters = Computation::without_witness(program).setup();
+        let computation = Computation::without_witness(program);
+
+        let rng = &mut rand_0_7::rngs::StdRng::from_seed(get_random_seed().unwrap());
+        let (pk, vk) = ArkGM17::<T::ArkEngine>::circuit_specific_setup(computation, rng).unwrap();
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/ark/gm17.rs:26]: Can T be inferred here?

File: zokrates_core/src/proof_system/ark/gm17.rs [hunk 5]
```diff
@@ -114,26 +116,24 @@ impl NonUniversalBackend<Bw6_761Field, GM17> for Ark {
     fn setup<I: IntoIterator<Item = Statement<Bw6_761Field>>>(
         program: ProgIterator<Bw6_761Field, I>,
     ) -> SetupKeypair<<GM17 as Scheme<Bw6_761Field>>::VerificationKey> {
-        let parameters = Computation::without_witness(program).setup();
+        let computation = Computation::without_witness(program);
+
+        let rng = &mut rand_0_7::rngs::StdRng::from_seed(get_random_seed().unwrap());
+        let (pk, vk) = ArkGM17::<BW6_761>::circuit_specific_setup(computation, rng).unwrap();
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/ark/gm17.rs:122]: Same here

File: zokrates_core/src/proof_system/ark/gm17.rs [hunk 6]
```diff
@@ -151,13 +151,15 @@ impl Backend<Bw6_761Field, GM17> for Ark {
...
-        let params =
+        let pk =
             ProvingKey::<<Bw6_761Field as ArkFieldExtensions>::ArkEngine>::deserialize_uncompressed(
                 &mut proving_key.as_slice(),
             )
                 .unwrap();
 
-        let proof = computation.prove(&params);
+        let rng = &mut rand_0_7::rngs::StdRng::from_seed(get_random_seed().unwrap());
+        let proof = ArkGM17::<BW6_761>::prove(&pk, computation, rng).unwrap();
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/ark/gm17.rs:161]: Same here

File: zokrates_core/src/proof_system/ark/groth16.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+        let computation = Computation::without_witness(program);
+
+        let rng = &mut rand_0_7::rngs::StdRng::from_seed(get_random_seed().unwrap());
+        let (pk, vk) = Groth16::<T::ArkEngine>::circuit_specific_setup(computation, rng).unwrap();
+
+        let mut pk_vec: Vec<u8> = Vec::new();
+        pk.serialize_uncompressed(&mut pk_vec).unwrap();
+
+        let vk = VerificationKey {
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/ark/groth16.rs:106]: There is a lot of code constructing these vks and proofs and it's used many times, is there a way to implement the From trait once and reuse it instead?

File: zokrates_core/src/proof_system/ark/mod.rs [hunk 8]
```diff
@@ -148,35 +141,20 @@ impl<T: Field + ArkFieldExtensions, I: IntoIterator<Item = Statement<T>>> ProgIt
...
-
-        let proof = create_random_proof(self, params, rng).unwrap();
-
-        let pvk = prepare_verifying_key(&params.vk);
-
-        assert!(verify_proof(&pvk, &proof, &public_inputs).unwrap());
-
-        proof
-    }
+pub fn get_random_seed() -> Result<[u8; 32], getrandom::Error> {
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/ark/mod.rs]: Can you explain why this change is required in this PR? Did the ark API change? If so, what did it use under the hood before for randomness?

File: zokrates_core/src/proof_system/mod.rs [hunk 9]
```diff
@@ -18,7 +18,12 @@ use serde::{Deserialize, Serialize};
 use std::io::{Read, Write};
 #[cfg(feature = "bellman")]
 use zokrates_field::BellmanFieldExtensions;
-use zokrates_field::Field;
+use zokrates_field::{Bls12_377Field, Bls12_381Field, Bn128Field, Field};
+
+pub trait NotBw6_761Field {}
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/mod.rs:23]: Shouldn't this happen in `zokrates_field` instead?

**Review by Schaeff — CHANGES_REQUESTED** [review]

File: zokrates_core/src/proof_system/ark/mod.rs (hunk 8 above)
Comment (dark64) [reply zokrates_core/src/proof_system/ark/mod.rs]: We used `from_entropy()` function from `rand 0.7` which does not work well with wasm. It compiles fine but you get a runtime error `target not supported`. Newest versions of rand `>0.8` use this `getrandom` crate under the hood.

File: zokrates_core/src/proof_system/mod.rs (hunk 9 above)
Comment (dark64) [reply zokrates_core/src/proof_system/mod.rs:23]: This thing is a workaround and a bit messy, so I wasn't sure where it belongs. Moving this to `zokrates_field` gives the following ```upstream crates may add a new impl of trait `zokrates_field::NotBw6_761Field` for type `zokrates_field::Bw6_761Field` in future versions```

File: zokrates_core/src/proof_system/ark/groth16.rs (hunk 7 above)
Comment (dark64) [reply zokrates_core/src/proof_system/ark/groth16.rs:106]: This will be done in another PR (regarding https://github.com/Zokrates/ZoKrates/issues/1079)

File: zokrates_core/src/proof_system/ark/gm17.rs (hunk 4 above)
Comment (dark64) [reply zokrates_core/src/proof_system/ark/gm17.rs:26]: not in this case

File: zokrates_core/src/proof_system/ark/mod.rs (hunk 8 above)
Comment (dark64) [reply zokrates_core/src/proof_system/ark/mod.rs]: Actually, after further research, I've found there is a feature in `rand 0.7` we can enable regarding wasm-bindgen, so I will revert these changes!

File: zokrates_core/src/proof_system/ark/groth16.rs [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+            gamma_abc: vk
+                .gamma_abc_g1
+                .iter()
+                .map(parse_g1::<Bw6_761Field>)
+                .collect(),
+        };
+
+        SetupKeypair::new(vk, pk_vec)
+    }
+}
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/ark/groth16.rs:214]: We should have some tests here, like in ark/gm17

**Review by Schaeff — APPROVED** [review]
